messaging/src/main/protos/Messages.proto

```diff
@@ -0,0 +1,185 @@
+}
+
+// Requires authentication
+//
+// Removes the recorded registration for the client's Address.
+message Unregister {}
+
+// Requires anonymous connection
+//
+// A request to retrieve preKey information for all registered devices for the given identityKey except those listed in knownDeviceIds.
+message RequestPreKeys {
+  bytes          identityKey    = 1;  // The identityKey for which to retrieve preKeys.
+  repeated bytes knownDeviceIds = 2;  // Devices of this identity which the client already knows about and doesn't need preKeys for.
+}
+
+// Information about a PreKey for a specific Address.
+message PreKey {
+  bytes deviceId       = 1;  // The deviceId that this key material belongs to
+  bytes signedPreKey   = 2;  // The most recent signedPreKey for the device at this Address. See https://crypto.stackexchange.com/questions/72148/signal-protocol-how-is-signed-preKey-created
+  bytes oneTimePreKey  = 3;  // One disposable preKey for the device at this Address. May be empty if none were available (that's okay, Signal can still do an X3DH key agreement without it).
+}
+
+// A list of PreKeys for all of an identityKey's devices, sent in response to RequestPreKeys
+message PreKeys {
+  repeated PreKey preKeys = 1;  // One or more preKeys
+}
+
+// A notification from the server to the client that we're running low on oneTimePreKeys for the Address associated to this connection.
+//
+// Clients may choose to respond to this by sending a Register message with some more preKeys. This does not have to be tied to the initial PreKeysLow message.
+message PreKeysLow {
+  uint32 keysRequested = 1;  // The number of additional oneTimePreKeys that the server is requesting.
+}
+
+// Requires anonymous connection
+//
+// A message from one client to another.
+message OutboundMessage {
+  Address to                        = 1;  // The Address of the message recipient
+  bytes   unidentifiedSenderMessage = 2;  // A sealed sender message (opaque to tassis). This is what will be delivered to the recipient.
+}
+
+// Requests attachment upload authorizations.
+message RequestUploadAuthorizations {
+  int32 numRequested = 1; // the number of authorizations requested. The server may not return the number requested.
+}
+
+// Provides authorization to upload an attachment to cloud storage
+message UploadAuthorization {
+  string              uploadURL              = 1; // The URL to which to upload
+  map<string, string> uploadFormData         = 2; // This form data needs to be included with the upload in order to authorize it
+  int64               authorizationExpiresAt = 3; // The unix timestamp in milliseconds when this authorization expires and can no longer be used
+  int64               maxUploadSize          = 4; // The maxmimum number of bytes that are allowed to be uploaded
+  string              downloadURL            = 5; // The URL from which the attachment may be downloaded once it has been uploaded
+}
```

Comment:
This is interesting.. would there be an account setting for this and users have the option to automatically sync attachments to different cloud storage providers (S3, Google Drive, etc.)?
If we let users authenticate with cloud providers via OAuth, that's something that could be done server-side possibly via the auth server. 
Have we thought about uploading attachments to our own servers (I guess Replica?) and handling copying the content to different cloud storage providers from there? It would be cool to also store the messages the attachments were sent with to give users an easy way to look them up later on. There was an app called Chatology recently discontinued that had search functionality like that: https://flexibits.com/chatology

Reply:
Durable attachments backup would be an interesting feature, and possibly a nice "pro" feature. The current attachments mechanism is there purely for purposes of facilitating transfer of attachments (via an S3-like service called Wasabi). We don't do that via Redis because it requires tons of memory/storage and would slow down the core message exchange, which we always want to keep as low latency as possible.

Reply:
> Have we thought about uploading attachments to our own servers

We didn't build it, but the attachments are stored on our own Wasabi account. See https://github.com/getlantern/tassis/tree/ox/simplifiedaddressing/attachments/s3attachments